plugin/cache_test.go

```diff
@@ -369,13 +373,92 @@ func (suite *CacheTestSuite) TestCachedListSetEntryID() {
 	}
 }
 
-func (suite *CacheTestSuite) TestCachedOpen() {
-	mockReader := strings.NewReader("foo")
-	suite.testCachedDefaultOp(OpenOp, "Open", mockReader, mockReader, func(ctx context.Context, e Entry) (interface{}, error) {
-		return cachedOpen(ctx, e.(Readable))
+func (suite *CacheTestSuite) TestCachedRead_DefaultOp() {
+	// This also tests a successful read of a ReadableCorePluginEntry
+	mockRawContent := []byte("some raw content")
+	mockContent := newEntryContent(mockRawContent)
+	suite.testCachedDefaultOp(ReadOp, "Read", mockRawContent, mockContent, func(ctx context.Context, e Entry) (interface{}, error) {
+		return cachedRead(ctx, e)
 	})
 }
 
+func (suite *CacheTestSuite) TestCachedRead_ReadableCorePluginEntry_ErroredRead() {
+	entry := newCacheTestsMockEntry("foo")
+	entry.DisableDefaultCaching()
+	entry.SetTestID("/foo")
+
+	ctx := context.Background()
+	expectedErr := fmt.Errorf("an error")
+	entry.On("Read", ctx).Return([]byte{}, expectedErr)
+
+	_, err := cachedRead(ctx, entry)
+	suite.Equal(expectedErr, err)
+}
+
+func (suite *CacheTestSuite) TestCachedRead_ReadableExternalPluginEntry_SuccessfulRead() {
+	entry := &externalPluginEntry{
+		EntryBase: NewEntry("foo"),
+		methods: map[string]methodInfo{
+			"read": methodInfo{signature: defaultSignature, result: "some raw content"},
+		},
+	}
+	entry.DisableDefaultCaching()
+	entry.SetTestID("/foo")
+
+	expectedContent := newEntryContent([]byte("some raw content"))
+
+	actualContent, err := cachedRead(context.Background(), entry)
+	if suite.NoError(err) {
+		suite.Equal(expectedContent, actualContent)
+	}
+}
+
+func (suite *CacheTestSuite) TestCachedRead_ReadableExternalPluginEntry_ErroredRead() {
+	entry := &externalPluginEntry{
+		EntryBase: NewEntry("foo"),
+		methods: map[string]methodInfo{
+			"read": methodInfo{signature: defaultSignature, result: []byte("invalid content")},
+		},
+	}
+	entry.DisableDefaultCaching()
+	entry.SetTestID("/foo")
+
+	_, err := cachedRead(context.Background(), entry)
+	suite.Regexp("Read.*string", err.Error())
+}
+
+type cacheTestsMockBlockReadableEntry struct {
+	*cacheTestsMockEntry
+}
+
+func (m *cacheTestsMockBlockReadableEntry) Read(ctx context.Context, size int64, offset int64) ([]byte, error) {
+	args := m.Called(ctx, size, offset)
+	return args.Get(0).([]byte), args.Error(1)
+}
+
+func (suite *CacheTestSuite) TestCachedRead_BlockReadableCorePluginEntry() {
+	entry := &cacheTestsMockBlockReadableEntry{
+		cacheTestsMockEntry: newCacheTestsMockEntry("foo"),
+	}
+	entry.DisableDefaultCaching()
+	entry.SetTestID("/foo")
+
+	ctx := context.Background()
+	expectedRawContent := []byte("some raw content")
+	entry.On("Read", ctx, int64(10), int64(0)).Return(expectedRawContent, nil).Once()
+
+	content, err := cachedRead(ctx, entry)
```

Comment:
I feel like somewhere in here we should be verifying that the length of content we got back matched the requested length.

Reply:
Or at least actualLength <= requestedLength. Because we're probably going to truncate anything longer than the requested length.

Reply:
Not sure what you mean. I ended up adding a test-case for `content.size`.

I'm thinking any `actualLength <= requestedLength` checks should be done in `plugin.Read`.

Reply:
Yes, I think you're right.

Reply:
My main observation in the test was that you requested a read of size 10, but get back content longer than that. And nothing fails.